### PR TITLE
random_circuit function

### PR DESCRIPTION
**Pull request by ajavadia**
Function for creating random quantum circuits (similar to existing functions for creating random unitaries, states, etc).
Replaces a previous function in the visualization tests folder.

### CODE REVIEW

**Comment by jaygambetta** [comment]
What is this random over

**Comment by ajavadia** [comment]
@jaygambetta This just builds up a circuit of desired shape by uniform-sampling from the list of possible gates. It is not Haar-random or anything. Our `random_state` and `random_unitary` functions are over Haar, so one may also use those then synthesize a circuit from them. But you will not have control over e.g. the circuit depth.

@mtreinish this is similar to the function you worked on for the random tests, but I think the depth in that one was not actual circuit depth. Also I inferred parameters from the gate signatures here, for compactness.

**Comment by jaygambetta** [comment]
I know but let’s define it so it’s clear. Can you talk to @adcorcol and make it the same as he is doing for the optimized and then can we use the same name. I think it’s hardware efficient

**Comment by jaygambetta** [comment]
Also since this is the first one can we put in a folder library

**Comment by mtreinish** [comment]
@ajavadia sure, this is similar enough and works basically the same way. Yeah, depth was the wrong name for that argument in my function. It wasn't actually circuit depth, but just a raw count of how many gates were added. I specifically avoided using signature inspection for speed reasons, especially since we already had to classify the gates by qubit count splitting the list by param counts was easy enough. Not that I think the overhead for this function is super important so signature inspection is probably fine.

File: test/python/circuit/test_random_circuit.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            self.assertIsInstance(nd.op, Measure)
+
+    def test_random_circuit_conditional_reset(self):
+        """Test generating random circuits with conditional and reset.
+        """
+        width = 1
+        depth = 100
+        circ = random_circuit(width, depth, conditional=True, reset=True)
+        self.assertEqual(circ.width(), 2*width)
+        self.assertIn('reset', circ.count_ops())
```
Comment (mtreinish) [inline test/python/circuit/test_random_circuit.py]: Is there actually a sure thing that we'll get a a circuit with a reset in it, isn't it randomly selected from the 1 qubit ops? I feel like this is just asking for non-deterministic failures in CI.

File: qiskit/circuit/random/utils.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+from qiskit.extensions import (IdGate, U0Gate, U1Gate, U2Gate, U3Gate, XGate,
+                               YGate, ZGate, HGate, SGate, SdgGate, TGate,
+                               TdgGate, RXGate, RYGate, RZGate, CnotGate,
+                               CyGate, CzGate, CHGate, CrzGate, Cu1Gate,
+                               Cu3Gate, SwapGate, RZZGate,
+                               ToffoliGate, FredkinGate)
+from qiskit.exceptions import QiskitError
+
+
+def random_circuit(width, depth, max_operands=3, measure=False,
```
Comment (nonhermitian) [inline qiskit/circuit/random/utils.py]: Can we not use `width` here, instead using `n_qubits` or `num_qubits`.  The `width` is not the number of qubits in a circuit with a classical register (due to the possibility of conditional operations).  This also cleans up the check that `circ.width() == 2*width` which is a bit confusing.

File: test/python/circuit/test_random_circuit.py (hunk 1 above)
Comment (ajavadia) [reply test/python/circuit/test_random_circuit.py]: ok I fixed the seed to make it deterministic.

File: qiskit/circuit/random/utils.py (hunk 2 above)
Comment (ajavadia) [reply qiskit/circuit/random/utils.py]: ok changed the arg to n_qubits.

File: qiskit/circuit/random/utils.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+def random_circuit(n_qubits, depth, max_operands=3, measure=False,
+                   conditional=False, reset=False, seed=None):
+    """Generate random circuit of arbitrary size and form.
+
+    Args:
+        n_qubits (int): number of quantum wires
+        depth (int): layers of operations (i.e. critical path length)
+        max_operands (int): maximum operands of each gate (between 1 and 3)
+        measure (bool): if True, measure all qubits at the end
+        conditional (bool): if True, insert middle measurements and conditionals
```
Comment (kdk) [inline qiskit/circuit/random/utils.py]: I don't think we end up generating middle measurements, but that's not blocking.

**Pull request merged**